Add all stage metrics to tools output

nartal1 opened the pull request:
This fixes https://github.com/NVIDIA/spark-rapids-tools/issues/928 and fixes https://github.com/NVIDIA/spark-rapids-tools/issues/1081

This PR adds a new output file(stage_level_all_metrics.csv) which has the information of all accumulables aggregated at stage level including Gpu accumulables(if any) in the eventlog. 

In some cases, there is just one entry in the eventlog for a given accumulator. In such cases, the min,median and max values are assigned default value of `0` and the actual value is in `total` column. This is to similar where we assign default values for sql_metrics_for_application.csv.

We drop those accumulators from output file if no metrics are generated.

The output format is as below:
```
appIndex,stageId,accumulatorId,name,min,median,max,total
1,"0",46,"duration",2694,4051,4051,6745
1,"0",48,"number of input batches",0,0,0,6
1,"10",1010,"gpuSemaphoreWait",0,1196,68449,11682104
1,"10",1013,"gpuSpillToDiskTime",229,2252,17319,1682175
1,"10",1015,"gpuReadSpillFromDiskTime",121,1790,9318,2034767
1,"10",1016,"gpuSplitAndRetryCount",0,0,0,1
1,"10",1018,"gpuSpillToHostTime",11,2087,15533,1783900
1,"10",1039,"output rows",177,476370,1376040,1015737409
1,"10",1040,"output columnar batches",177,61421,143354,118334405
```
 Added unit test to simulate eventlog containing GPU metrics and capture them.

<!--

Thank you for contributing to Tools of RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

nartal1 commented:
@amahussein @tgravescs - Would be good to know if the information in the output file is okay OR any columns need to be added/removed. Thanks!

amahussein commented:
> > This PR adds a new output file(gpu_metrics_for_application.csv) which has the information of Gpu accumulables(if any) in the eventlog.
> 
> Do we want to have a separate CSV file for GPU metrics. This is a stageLevel accumulator view. If we name it to something more generic, then it can be used to dump all other accumulators. CC: @tgravescs Do you have any preferences?



> > This PR adds a new output file(gpu_metrics_for_application.csv) which has the information of Gpu accumulables(if any) in the eventlog.
> 
> Do we want to have a separate CSV file for GPU metrics. This is a stageLevel accumulator view. If we name it to something more generic, then it can be used to dump all other accumulators. CC: @tgravescs Do you have any preferences?

I discussed with Tom offline.
Let's remove the filter and make the file generic for all the Accumulators CPU/GPU.

- One main fix to do in this PR is to generate the view by the Qualification tool as well. 
- this PR will also close issue 1081.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala:
```diff
@@ -342,6 +342,50 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
+  def generateGpuAccums(): Seq[GpuAccumProfileResults] = {
+    def isGpuAccum(accum: TaskStageAccumCase): Boolean = {
+      accum.name.exists(_.startsWith(ToolUtils.GpuPrefix))
+    }
```
Removed this code based on offline conversation since we are dumping all metrics to output file and not just gpu specific metrics.

nartal1 commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala:
```diff
@@ -342,6 +342,50 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
+    val groupedByAccumulatorId = gpuAccums.groupBy(_.accumulatorId)
+    groupedByAccumulatorId.flatMap { case (accumulatorId, accums) =>
+      // Extract and sort the update values, defaulting to 0 if not present
+      val sortedUpdates = accums.flatMap(_.update).toSeq.sorted
```
This is to sort the list to get the min, median and max for a given accumulator. The other sort is for the View.

amahussein replied:
I see.
It looks like a code efficiency problem. Picking min, max, med, and then sum seems to be done in several iterations than necessary.
Anyway, this is not part of this PR.

amahussein replied:
FYI @bilalbari

nartal1 commented on core/src/test/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationInfoSuite.scala:
```diff
@@ -202,6 +203,51 @@ class ApplicationInfoSuite extends FunSuite with Logging {
+      TrampolineUtil.withTempDir { tmpEventLogDir =>
+        val eventLogFilePath = Paths.get(tmpEventLogDir.getAbsolutePath, "gpu_metrics_eventlog")
+        // scalastyle:off line.size.limit
+        val eventLogContent =
```
The current eventlogs in the tests were generated long ago and are quite old. They don't include any gpu metrics. I have filed a follow on issue https://github.com/NVIDIA/spark-rapids-tools/issues/1154 to discuss updating of eventlogs for unit-tests

amahussein reviewed (commented):
The PR's description needs to be updated

For convenience, I am attaching sample output. [stage_level_aggregated_task_metrics.csv](https://github.com/user-attachments/files/16058178/stage_level_aggregated_task_metrics.csv)

I can see that we have internal accumulators.
I believe that `internal.metrics.*` are basically duplicate of the `stage_level_aggregated_task_metrics.csv` 

**A couple of questions:**

- Do we want to list `internal.metrics.*` in this file? A concern is redundancy and that having same information in multiple files might be confusing.
- Should we drop "zero" rows to reduce the size of the output (rows that have 0-value in all the columns)?

CC: @tgravescs

amahussein reviewed (changes requested):
I triggered qualification tool. The new file does not get generated under `raw_metrics` subdirectory

nartal1 commented:
> I triggered qualification tool. The new file does not get generated under raw_metrics subdirectory
 
Thanks @amahussein ! I forgot  that we have another path to generate the output files in raw_metrics. Updated the code to generate the csv file under raw_metrics directory. PTAL.

> Should we drop "zero" rows to reduce the size of the output (rows that have 0-value in all the columns)?

I have updated the code to remove those rows where metrics are not updated at all i.e value of all columns is zero. Will revert it if we want those rows as well.

> Do we want to list internal.metrics.* in this file? A concern is redundancy and that having same information in multiple files might be confusing.

I was not certain if we wanted to remove it since the issue was to dump all the metrics to a output file. @tgravescs - Could you please let us know if we need to keep internal.* ones ? Also, is it okay to remove rows from the output file if all the values are "zeros"?

amahussein commented:
Thanks @nartal1 
I discussed offline with Tom.

For this PR, we can dump everything without filtering any rows. This means:

- the rows with all zero-values should be dumped out as well.
- the internal metrics should show up too.

Later, I can investigate with Bilal the performance impact of re-aggregating the internal Spark metrics as part of figuring out low-hanging fruits for https://github.com/NVIDIA/spark-rapids-tools/issues/367

amahussein commented:
Also, let's change the name from `stage_level_all_metrics_for_application.csv` to `stage_level_all_metrics.csv`

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ApplicationSummaryInfo.scala:
```diff
@@ -26,6 +26,7 @@ case class ApplicationSummaryInfo(
     rapidsProps: Seq[RapidsPropertyProfileResult],
     rapidsJar: Seq[RapidsJarProfileResult],
     sqlMetrics: Seq[SQLAccumProfileResults],
+    gpuMetrics: Seq[AccumProfileResults],
```
not `gpuMetrics` anymore.
Also, this is very large array. I am not sure if this is necessary to store this as a field in memory and keep it around.
If `generateStageLevelAccums` is defined in `AppSQLPlanAnalyzer`. Then, it could be called to generate the report without storing it as an extra field in ProfileResult.
Unless, we need this somewhere else.

nartal1 replied:
Thanks @amahussein ! renamed gpuMetrics to stageMetrics. As discussed with you offline, the code path for generating Qualification output files and Profile output files are different where the Profile outputwriter doesn't take ApplicationInfo as an argument. So keeping it as it is for now.

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala:
```diff
@@ -342,6 +342,59 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
+   *
+   * @return a sequence of AccumProfileResults
+   */
+
```
empty line

nartal1 replied:
done

amahussein commented on core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala:
```diff
@@ -342,6 +342,59 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
+   * @return a sequence of AccumProfileResults
+   */
+
+  def generateStageLevelAccums(): Seq[AccumProfileResults] = {
```
nit: this is actually a stageLevel thing but it is defined in the AppSQLPlanAnalyzer which does not look like a good fit.
I am fine to keep it here for now.

amahussein commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/ToolUtils.scala:
```diff
@@ -43,6 +43,8 @@ object ToolUtils extends Logging {
     "spark.repl.local.jars")
   val RAPIDS_JAR_REGEX = "(.*rapids-4-spark.*jar)|(.*cudf.*jar)".r
 
+  val GpuPrefix: String = "gpu"
```
not needed

nartal1 replied:
removed

amahussein reviewed (approved):
Thanks @nartal1 
LGTME
